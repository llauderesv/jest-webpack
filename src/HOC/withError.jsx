import React from 'react';
import ErrorMessage from '../components/ErrorMessage';
import Button from 'react-bootstrap/Button';

const withError = WrappedComponent => ({ error, fetchEvent, ...restProps }) => {
  return error ? (
    <ErrorMessage message={error.message}>
      <Button variant="danger" onClick={fetchEvent}>
        Retry
      </Button>
    </ErrorMessage>
  ) : (
    <WrappedComponent {...restProps} />
  );
};

export default withError;
