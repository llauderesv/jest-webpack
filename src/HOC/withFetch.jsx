import React, { useEffect, useState } from 'react';
import request from '../api/axios-config';

const delay = ms =>
  new Promise((resolve, reject) => {
    setTimeout(() => {
      const rnd = Math.ceil(Math.random() * 50);

      if (rnd > 30)
        reject({ message: 'There was something wrong in the server.' });
      resolve({
        data:
          'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Perferendis iusto at est hic excepturi! Tempore sapiente delectus iste, excepturi deserunt, magnam mollitia doloremque corrupti suscipit nesciunt reiciendis, sed nulla aut.',
      });
    }, ms);
  });

export default function withFetch(WrappedComponent) {
  function fetch({ ...restProps }) {
    const data = useFetchUsers();

    return <WrappedComponent {...data} {...restProps} />;
  }

  return fetch;
}

function useFetchUsers() {
  const defaultState = {
    data: null,
    isLoading: false,
    error: null,
  };
  const [data, setData] = useState(defaultState);

  function fetchUsers() {
    setData({ ...data, isLoading: true });

    delay(3000)
      .then(response => {
        setData({
          ...data,
          data: response.data,
          isLoading: false,
        });
      })
      .catch(error => {
        setData({
          ...data,
          error: error,
          isLoading: false,
          fetchEvent: fetchUsers,
        });
      });
  }

  useEffect(() => {
    fetchUsers();
  }, []);

  return data;
}
