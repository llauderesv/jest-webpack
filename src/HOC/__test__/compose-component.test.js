import React from 'react';

import withError from '../withError';
import withLoading from '../withLoader';

import TestRenderer from 'react-test-renderer';
import { compose } from 'redux';

import withFetch from '../withFetch';

const HelloWorld = _ => <h1>Hello World</h1>;

describe('Higher Order Component test suite.', () => {
  it('Should render withError component.', () => {
    const enhance = compose(
      withFetch,
      withError,
      withLoading,
    );

    // HelloWorld -> withLoading -> withError -> withFetch
    /**
     * withLoading Component Receives HelloWorld Component
     * withError Component Receives withLoading Component
     * withFetch Component Receives withError Component.
     *
     * withFetch Component Receives HelloWorld Component
     *
     */

    const EnhancedHelloWorld = enhance(HelloWorld);
    const app = TestRenderer.create(<EnhancedHelloWorld testProps={'test'} />);

    const tree = app.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
