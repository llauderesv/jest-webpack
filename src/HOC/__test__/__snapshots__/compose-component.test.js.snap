// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Higher Order Component test suite. Should render withError component. 1`] = `
<h1>
  Hello World
</h1>
`;
